### PR TITLE
Fix get reward info

### PR DESCRIPTION
**Pull request by pichaoqkc**
Return correct reward info without updating price

### CODE REVIEW

File: contracts/UniversalV3Staker.sol [hunk 1]
```diff
@@ -315,16 +315,32 @@ contract UniversalV3Staker is IUniversalV3Staker, Multicall {
...
+        uint256 timestamp = block.timestamp;
+
+        uint256 calculatedRewards = rewardCalc.getRewards(rewardUpdatedAt + 1, timestamp);
+        if (calculatedRewards == 0) return (0, 0);
+
+        uint256 rewardShareX64 = (calculatedRewards << 64) / uint256(liquidity);
+
+        if (currentTick >= deposit.tickLower && currentTick <= deposit.tickUpper) {
+            uint256 calculatedRewards = rewardCalc.getRewards(rewardUpdatedAt + 1, timestamp);
+            reward += calculatedRewards;
```
Comment (qizhou) [inline contracts/UniversalV3Staker.sol]: should be

reward += (rewareShareX64 * deposit.liquidity) >> 64?

File: contracts/UniversalV3Staker.sol [hunk 2]
```diff
@@ -315,16 +315,32 @@ contract UniversalV3Staker is IUniversalV3Staker, Multicall {
...
 
         uint24 tickLowerShifted = uint24(deposit.tickLower - TickMath.MIN_TICK + 1);
         uint24 tickUpperShifted = uint24(deposit.tickUpper - TickMath.MIN_TICK + 1);
         uint256 latestReward = _calculateReward(liquidity, tickLowerShifted, tickUpperShifted);
         reward = latestReward.sub(rewardDebt);
+
+        uint256 timestamp = block.timestamp;
+
+        uint256 calculatedRewards = rewardCalc.getRewards(rewardUpdatedAt + 1, timestamp);
+        if (calculatedRewards == 0) return (0, 0);
```
Comment (ninjaahhh) [inline contracts/UniversalV3Staker.sol]: is this right?

File: contracts/UniversalV3Staker.sol [hunk 3]
```diff
@@ -332,13 +333,9 @@ contract UniversalV3Staker is IUniversalV3Staker, Multicall {
...
         uint256 calculatedRewards = rewardCalc.getRewards(rewardUpdatedAt + 1, timestamp);
-        if (calculatedRewards == 0) return (0, 0);
-
-        uint256 rewardShareX64 = (calculatedRewards << 64) / uint256(liquidity);
 
-        if (currentTick >= deposit.tickLower && currentTick <= deposit.tickUpper) {
-            uint256 calculatedRewards = rewardCalc.getRewards(rewardUpdatedAt + 1, timestamp);
-            reward += calculatedRewards;
+        if (currentTick >= tickLowerShifted && currentTick <= tickUpperShifted) {
+            reward += calculatedRewards ;
```
Comment (ninjaahhh) [inline contracts/UniversalV3Staker.sol]: safe math

File: contracts/UniversalV3Staker.sol [hunk 4]
```diff
@@ -315,16 +315,29 @@ contract UniversalV3Staker is IUniversalV3Staker, Multicall {
...
         uint24 tickLowerShifted = uint24(deposit.tickLower - TickMath.MIN_TICK + 1);
         uint24 tickUpperShifted = uint24(deposit.tickUpper - TickMath.MIN_TICK + 1);
         uint256 latestReward = _calculateReward(liquidity, tickLowerShifted, tickUpperShifted);
         reward = latestReward.sub(rewardDebt);
+
+        uint256 timestamp = block.timestamp;
+
+        uint256 calculatedRewards = rewardCalc.getRewards(rewardUpdatedAt + 1, timestamp);
+
+        if (currentTick >= tickLowerShifted && currentTick <= tickUpperShifted) {
```
Comment (qizhou) [inline contracts/UniversalV3Staker.sol:333]: Both timestampe and calculatedRewards can be moved into if () { }?

**Review by qizhou — APPROVED** [review]

File: contracts/UniversalV3Staker.sol [hunk 5]
```diff
@@ -314,17 +314,33 @@ contract UniversalV3Staker is IUniversalV3Staker, Multicall {
...
+            (, liquidity, rewardDebt) = stakes(tokenId, UniversalIncentiveId.compute(key));
+            require(liquidity > 0, 'UniswapV3Staker::getRewardInfo: stake does not exist');
+            tickLowerShifted = uint24(deposit.tickLower - TickMath.MIN_TICK + 1);
+            tickUpperShifted = uint24(deposit.tickUpper - TickMath.MIN_TICK + 1);
+            reward = _calculateReward(liquidity, tickLowerShifted, tickUpperShifted).sub(rewardDebt);
+        }
 
-        Deposit memory deposit = deposits[tokenId];
+        if (currentTick >= tickLowerShifted && currentTick <= tickUpperShifted) {
+            uint24 tickBeforeUpdate = uint24(lastTick - TickMath.MIN_TICK + 1);
```
Comment (ninjaahhh) [inline contracts/UniversalV3Staker.sol]:
```suggestion
            uint24 lastTickShifted = uint24(lastTick - TickMath.MIN_TICK + 1);
```

File: contracts/UniversalV3Staker.sol [hunk 6]
```diff
@@ -314,17 +314,33 @@ contract UniversalV3Staker is IUniversalV3Staker, Multicall {
...
+            uint208 liquidityLower = _cumulativeLiquidityLower.get(_cfNbits, tickBeforeUpdate);
+            uint208 liquidityUpper = _cumulativeLiquidityUpper.get(_cfNbits, tickBeforeUpdate);
+            uint208 totalLiq = liquidityLower - liquidityUpper;
```
Comment (ninjaahhh) [inline contracts/UniversalV3Staker.sol:337]: assert no overflow?

File: contracts/UniversalV3Staker.sol [hunk 7]
```diff
@@ -314,17 +314,34 @@ contract UniversalV3Staker is IUniversalV3Staker, Multicall {
...
 
-        Deposit memory deposit = deposits[tokenId];
+        if (currentTick >= tickLowerShifted && currentTick <= tickUpperShifted) {
+            uint24 lastTickShifted = uint24(lastTick - TickMath.MIN_TICK + 1);
+            uint208 liquidityLower = _cumulativeLiquidityLower.get(_cfNbits, lastTickShifted);
+            uint208 liquidityUpper = _cumulativeLiquidityUpper.get(_cfNbits, lastTickShifted);
+            uint208 totalLiq = liquidityLower - liquidityUpper;
+            require(totalLiq <= liquidityLower, 'UniswapV3Staker::gerRewardInfo: overflow');
+            uint256 calculatedRewards = key.rewardCalc.getRewards(rewardUpdatedAt + 1, block.timestamp);
+            uint256 rewardShareX64 = (calculatedRewards.mul(2**64)).div(uint256(totalLiq));
```
Comment (ninjaahhh) [inline contracts/UniversalV3Staker.sol]:
```suggestion
            uint256 rewardShareX64 = calculatedRewards.mul(2**64).div(uint256(totalLiq));
```

File: contracts/UniversalV3Staker.sol [hunk 8]
```diff
@@ -314,17 +314,34 @@ contract UniversalV3Staker is IUniversalV3Staker, Multicall {
...
+            uint256 rewardX64 = uint256(liquidity).mul(uint256(rewardShareX64));
```
Comment (ninjaahhh) [inline contracts/UniversalV3Staker.sol]:
```suggestion
            uint256 rewardX64 = uint256(liquidity).mul(rewardShareX64);
```

File: contracts/UniversalV3Staker.sol (hunk 7 above)
Comment (ninjaahhh) [reply contracts/UniversalV3Staker.sol]: also update similar code in another place

**Review by ninjaahhh — APPROVED** [review]

**Pull request merged**